### PR TITLE
Add AdditiveNoiseModel (ANM) simulator dataset

### LINKED ISSUE #3418: [ENH] Add AdditiveNoiseModel (ANM) simulator dataset
**Is your feature request related to a problem? Please describe.**

Additive Noise Models (Hoyer et al., 2008) are a standard benchmark for causal discovery methods. Currently there is no way to generate ANM data through `load_dataset()`. Users have to build these models manually or use external packages.

Part of #3402. Depends on #3404.

**Describe the solution you'd like**

New file at `pgmpy/datasets/anm.py`:

The `AdditiveNoiseModel` class inherits from `_SimulationMixin` and `_BaseDataset`. It generates a random DAG, assigns non-linear functions (e.g. polynomial, GP) to each edge, and adds independent noise. The simulator accepts `n_nodes`, `edge_prob`, `noise_type`, and `function_type` as kwargs, with `seed` controlling both graph generation and data sampling for reproducibility. `load_ground_truth()` returns the true causal DAG used during generation.

**Describe alternatives you've considered**

Using `sempler` as a backend. `sempler` is already a test dependency in pgmpy — worth checking whether it can be used internally or if a from-scratch implementation is cleaner.

**Additional context**

- Design reference: [enhancement_proposals#3](https://github.com/pgmpy/enhancement_proposals/pull/3)
- Paper: Hoyer et al., "Nonlinear causal discovery with additive noise models", NeurIPS 2008

**Tests**

- [x] #3419

### LINKED ISSUE #3419: [TEST] Tests for AdditiveNoiseModel simulator
Add test coverage for the `AdditiveNoiseModel` simulator introduced in the parent issue.

Parent issue: #3418 

**Test plan**

The tests verify that `load_dataset("anm", ...)` routes correctly through `load_dataset()` and returns a `Dataset` with the expected DataFrame shape and a valid `DAG` as ground truth. Seed reproducibility is checked by loading twice with the same seed and asserting identical output. Different `function_type` values are tested to confirm they all produce valid data without errors. Finally, `"anm"` is verified to appear in `list_datasets(is_simulated=True)` for discoverability.

**Verification**

```bash
pytest -v pgmpy/tests/test_datasets/test_datasets.py
```

### PR DESCRIPTION
**Pull request by Gitanaskhan26**
**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [x] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md#ai-usage-policy)?
- [x] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.

- Please list the AI tool(s) used, along with the model and its version used.

[Please replace this with your answer]

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.

[Please replace this with your answer]

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?

[Please replace this with your answer]

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

[Please replace this with your answer]

### Issue number(s) that this pull request fixes
- Fixes #3418
- Fixes #3419 

### List of changes to the codebase in this pull request

- **New file: `pgmpy/datasets/anm.py`** — Adds the `AdditiveNoiseModel` class inheriting from `BaseSimulatedDataset`. Implements a from-scratch Additive Noise Model simulator that generates data as X_j = f(Pa(X_j)) + N_j.

  Key features:
  - **Graph input**: Accepts a user-provided `pgmpy.base.DAG` or generates a random one via `DAG.get_random(n_nodes, edge_prob, seed)`.
  - **Non-linearity presets**: Three built-in function types — `sine_add`, `polynomial`, `sigmoid_add` — with randomly sampled coefficients from a configurable `weight_range`.
  - **Custom callables**: Users can pass any Python callable as `function_type` for full flexibility.
  - **Noise distributions**: Default Gaussian noise via numpy. Optionally accepts `skpro` distribution objects (soft dependency) for custom noise terms.
  - **Input validation**: Clear error messages for invalid `function_type`, bad `weight_range`, conflicting `noise_scale` + custom noise, non-DAG graphs, etc.
  - **Reproducibility**: Seeded via `np.random.default_rng(seed)` for the default path.

### Test Coverage

Implementation is tracked separately in #3419 . I've added the test cases to this PR as mentioned below.

Added `test_load_anm_dataset` in `pgmpy/tests/test_datasets/test_anm.py` to check the main happy paths. It makes sure the simulator works out of the box with default settings, handles custom sizes for nodes and samples, gives the exact same results when using the same seed, and keeps the graph structure independent of the math functions. It also verifies all three built-in presets, checks behavior when there are no connections, confirms the proper warning when a custom graph overrides other settings, accepts custom Python functions, handles single-node graphs, shows that noise levels change as expected, and ensures the dataset shows up correctly in the registry.

Added `test_anm_validation` in `pgmpy/tests/test_datasets/test_anm.py` to make sure the simulator catches mistakes early. It checks that the code raises clear errors for invalid function names, passing in non-function types, broken or negative weight ranges, requesting zero nodes, using impossible connection probabilities, setting negative noise scales, and asking for zero data points.

Added `test_anm_skpro_noise` in `pgmpy/tests/test_datasets/test_anm.py` to test the optional `skpro` library path. This test only runs if `skpro` is installed. It verifies that custom noise distributions like `Laplace` work correctly, blocks users from mixing custom noise with the `noise_scale` setting, and confirms that isolated variables stay statistically independent even when using non-Gaussian noise.

### Run command
```bash
pytest -v pgmpy/tests/test_datasets/test_anm.py

### CODE REVIEW

File: pgmpy/datasets/anm.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+      ``"sine_add"``, ``"polynomial"``, ``"sigmoid_add"``.  A callable must
+      accept an array of shape ``(n_samples, n_parents)`` and return an
+      array of shape ``(n_samples,)``.
+    * **noise_scale** (float, default 1.0) – Multiplier for the default
+      Gaussian noise.  Cannot be used together with a custom ``noise``
+      distribution.
+    * **weight_range** (tuple of float, default ``(0.5, 2.0)``) – Range
+      for randomly sampled coefficients in preset functions.  Ignored when
+      ``function_type`` is a callable.
+
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: Numpy style docstrings please.

File: pgmpy/datasets/anm.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+from pgmpy.base import DAG
+from pgmpy.datasets._base import BaseSimulatedDataset
+
+_SUPPORTED_FUNCTION_TYPES = {"sine_add", "polynomial", "sigmoid_add"}
+
+
+def _sigmoid(x: np.ndarray) -> np.ndarray:
+    """Numerically stable sigmoid."""
+    return np.where(x >= 0, 1 / (1 + np.exp(-x)), np.exp(x) / (1 + np.exp(x)))
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: Could such functions go in a separate class or directory?

File: pgmpy/datasets/anm.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    def _validate_params(
+        n_samples: int | None,
+        n_nodes: int,
+        edge_prob: float,
+        function_type: str | Any,
+        weight_range: tuple[float, float],
+        noise: Any,
+        noise_scale: float,
+    ) -> None:
+        """Validate simulator parameters, raising on invalid combinations."""
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: Could you mention here what all is getting validated here?

File: pgmpy/datasets/anm.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        Returns
+        -------
+        numpy.ndarray of shape (n_samples,)
+        """
+        if noise is None:
+            return rng.normal(0, 1, size=n_samples) * noise_scale
+
+        # Custom skpro distribution path
+        if noise_seed is not None:
+            np.random.seed(noise_seed)  # noqa: NPY002
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: does this raise a linter warning? why are you using noqa?

File: pgmpy/datasets/anm.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+      Gaussian noise.  Cannot be used together with a custom ``noise``
+      distribution.
+    * **weight_range** (tuple of float, default ``(0.5, 2.0)``) – Range
+      for randomly sampled coefficients in preset functions.  Ignored when
+      ``function_type`` is a callable.
+
+    References
+    ----------
+    Hoyer, P., Janzing, D., Mooij, J. M., Peters, J., & Schölkopf, B. (2008).
+    Nonlinear causal discovery with additive noise models. *NeurIPS*.
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: numpy style references please

File: pgmpy/datasets/anm.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    Simulated dataset from a random Additive Noise Model.
+
+    Generates data according to the model:
+
+    .. math::
+
+        X_j = f_j(\\text{Pa}(X_j)) + N_j
+
+    where :math:`f_j` are non-linear functions and :math:`N_j` are independent
+    noise terms (Hoyer et al., 2008).
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: could you check if this renders correct.

File: pgmpy/datasets/anm.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+
+def _sigmoid(x: np.ndarray) -> np.ndarray:
+    """Numerically stable sigmoid."""
+    return np.where(x >= 0, 1 / (1 + np.exp(-x)), np.exp(x) / (1 + np.exp(x)))
+
+
+class AdditiveNoiseModel(BaseSimulatedDataset):
+    """
+    Simulated dataset from a random Additive Noise Model.
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: might be a good idea to cite the reference here

File: pgmpy/datasets/anm.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+    _tags = {
+        "name": "anm",
+        "has_ground_truth": True,
+        "is_continuous": True,
+    }
+
+    # ------------------------------------------------------------------
+    # Private helpers
+    # ------------------------------------------------------------------
```
Comment (DARHWOLF) [inline pgmpy/datasets/anm.py]: I dont think we need these

File: pgmpy/datasets/anm.py (hunk 2 above)
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: sigmoid is a smooth function defined for x = (- \inf, \inf). What do we need to check `x >= 0`?

File: pgmpy/datasets/anm.py
Comment (ankurankan) [inline pgmpy/datasets/anm.py:1]: Set editor setting for line length = 120

File: pgmpy/datasets/anm.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+      non-linear function applied to parent values.  Supported strings:
+      ``"sine_add"``, ``"polynomial"``, ``"sigmoid_add"``.  A callable must
+      accept an array of shape ``(n_samples, n_parents)`` and return an
+      array of shape ``(n_samples,)``.
+    * **noise_scale** (float, default 1.0) – Multiplier for the default
+      Gaussian noise.  Cannot be used together with a custom ``noise``
+      distribution.
+    * **weight_range** (tuple of float, default ``(0.5, 2.0)``) – Range
+      for randomly sampled coefficients in preset functions.  Ignored when
+      ``function_type`` is a callable.
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: numpy doc style. Each parameter should be on a new line starting with `-`. Check numpydoc documentation for details.

File: pgmpy/datasets/anm.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+
+    References
+    ----------
+    Hoyer, P., Janzing, D., Mooij, J. M., Peters, J., & Schölkopf, B. (2008).
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Not the correct reference style either. Reference bibliography goes into `reference.bib`

File: pgmpy/datasets/anm.py [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+                raise ImportError(
+                    "skpro is required for custom noise distributions. Install with: pip install skpro"
+                ) from None
+            if not isinstance(noise, BaseDistribution):
+                raise TypeError(f"noise must be a skpro BaseDistribution instance or None, got {type(noise).__name__}.")
+
+    # ------------------------------------------------------------------
+    # Public API
+    # ------------------------------------------------------------------
+
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Remove this

File: pgmpy/datasets/anm.py [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+    @classmethod
+    def load_dataframe(
+        cls,
+        n_samples: int | None = None,
+        seed: int | None = None,
+        dag: DAG | None = None,
+        n_nodes: int = 5,
+        edge_prob: float = 0.5,
+        noise: Any = None,
+        function_type: str | Any = "sine_add",
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: This should be any set of callables, e.g., np.sin, np.cos. We can randomly sample from that and use those.

File: pgmpy/datasets/anm.py [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+        noise_scale: float = 1.0,
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Don't need `noise_scale`; this should come in the object specified in `noise`.

File: pgmpy/datasets/anm.py [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+            ``np.random.seed``).  Must be unique per node to ensure
+            independent noise draws.
+
+        Returns
+        -------
+        numpy.ndarray of shape (n_samples,)
+        """
+        if noise is None:
+            return rng.normal(0, 1, size=n_samples) * noise_scale
+
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: This needs to be specified in the documentation that you default to normal noise

File: pgmpy/datasets/anm.py [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+        # skpro distribution path — save/restore global RNG state so
+        # the simulator never leaks side effects to the caller.
+        state = np.random.get_state()  # noqa: NPY002
+        try:
+            if noise_seed is not None:
+                np.random.seed(noise_seed)  # noqa: NPY002
+            samples = noise.sample(n_samples=n_samples)
+        finally:
+            np.random.set_state(state)  # noqa: NPY002
+        return np.asarray(samples).flatten()
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: try to not use `try-except` or `try-finally` blocks.

File: pgmpy/datasets/anm.py [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+            ``function_type``, ``noise_scale``, and ``weight_range``.
+
+        Returns
+        -------
+        pgmpy.base.DAG
+        """
+        graph = cls._get_dag(dag, n_nodes, edge_prob, seed)
+        gt = DAG()
+        gt.add_nodes_from(graph.nodes())
+        gt.add_edges_from(graph.edges())
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Use the `self.model` pattern as we talked in the other PR

File: pgmpy/datasets/anm.py [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+      used.  ``skpro`` must be installed separately.
+    * **function_type** (str or callable, default ``"sine_add"``) – The
+      non-linear function applied to parent values.  Supported strings:
+      ``"sine_add"``, ``"polynomial"``, ``"sigmoid_add"``.  A callable must
+      accept an array of shape ``(n_samples, n_parents)`` and return an
+      array of shape ``(n_samples,)``.
+    * **noise_scale** (float, default 1.0) – Multiplier for the default
+      Gaussian noise.  Cannot be used together with a custom ``noise``
+      distribution.
+    * **weight_range** (tuple of float, default ``(0.5, 2.0)``) – Range
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Use default range as (-1, 1)

File: pgmpy/datasets/anm.py [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+            return lambda x, w=weights: (w * np.sin(x)).sum(axis=1)
+        elif function_type == "polynomial":
+            degrees = rng.choice([2, 3], size=n_parents)
+            return lambda x, w=weights, d=degrees: (w * np.power(x, d)).sum(axis=1)
+        elif function_type == "sigmoid_add":
+            return lambda x, w=weights: (w * _sigmoid(x)).sum(axis=1)
+        else:
+            raise ValueError(
+                f"Unknown function_type '{function_type}'. Supported: {sorted(_SUPPORTED_FUNCTION_TYPES)}."
+            )
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Once you start accepting the `function_type` as function objects, we won't need to do this if-else, we can simply do `(function_type(x))`

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Review by DARHWOLF — DISMISSED** [review]

File: pgmpy/datasets/anm.py [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+            from skpro.distributions.base import BaseDistribution
+
+            if not isinstance(noise, BaseDistribution):
+                raise TypeError(f"noise must be a skpro BaseDistribution instance or None, got {type(noise).__name__}.")
+
+    @classmethod
+    def load_dataframe(
+        cls,
+        n_samples: int | None = None,
+        seed: int | None = None,
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: `seed` should usually be the last argument.

File: pgmpy/datasets/anm.py [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+    Generates data according to the model:
+
+    .. math::
+
+        X_j = f_j(\\text{Pa}(X_j)) + N_j
+
+    where :math:`f_j` are non-linear functions and :math:`N_j` are
+    independent noise terms :cite:p:`hoyer_2008`.
+
+    Parameters accepted via keyword arguments in :func:`load_dataset`:
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Shouldn't have this line. This would not make sense once the online docs are rendered as `Parameters` will be rendered in its own separate section.

File: pgmpy/datasets/anm.py
Comment (ankurankan) [reply pgmpy/datasets/anm.py:1]: @Gitanaskhan26 Please modify your line length to 120 char.

File: pgmpy/datasets/anm.py [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+class AdditiveNoiseModel(BaseSimulatedDataset):
+    """Simulated dataset from a random Additive Noise Model.
+
+    Generates data according to the model:
+
+    .. math::
+
+        X_j = f_j(\\text{Pa}(X_j)) + N_j
+
+    where :math:`f_j` are non-linear functions and :math:`N_j` are
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: define what `Pa(X_j)` is.

File: pgmpy/datasets/anm.py [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+        ``edge_prob`` are ignored with a warning.
+    n_nodes : int, optional (default 5)
+        Number of variables in the random DAG.
+    edge_prob : float, optional (default 0.5)
+        Probability of an edge between any two topologically
+        ordered nodes in the random DAG.
+    noise : BaseDistribution, optional
+        A ``skpro`` distribution instance for noise terms.
+        When ``None``, standard normal :math:`\\mathcal{N}(0, 1)`
+        noise is used.  ``skpro`` must be installed separately.
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Needs to be either skpro or scipy distributions. Anything that we can sample from.

File: pgmpy/datasets/anm.py [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+from typing import Any
+
+import networkx as nx
+import numpy as np
+import pandas as pd
+
+from pgmpy.base import DAG
+from pgmpy.datasets._base import BaseSimulatedDataset
+
+_DEFAULT_FUNCTIONS = (np.sin, np.cos, np.tanh)
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Remove this. The default values should be in the function/method definition itself. This makes the code harder to read: 1. It doesn't really provide any abstraction, 2. We will have to jump from the method definition line to here to see what the default values are.

File: pgmpy/datasets/anm.py [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+
+        Returns
+        -------
+        DAG
+        """
+        if dag is not None:
+            if not isinstance(dag, DAG):
+                raise TypeError(f"dag must be a pgmpy.base.DAG instance, got {type(dag).__name__}.")
+            if not nx.is_directed_acyclic_graph(dag):
+                raise ValueError("dag must be an acyclic directed graph.")
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: DAG already checks for this. The DAG object is always valid.

File: pgmpy/datasets/anm.py [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+                    "dag was provided; n_nodes and edge_prob are ignored.",
+                    UserWarning,
+                    stacklevel=4,
+                )
+            return dag
+
+        if not isinstance(n_nodes, int) or n_nodes < 1:
+            raise ValueError(f"n_nodes must be a positive integer, got {n_nodes!r}.")
+        if not 0 <= edge_prob <= 1:
+            raise ValueError(f"edge_prob must be between 0 and 1, got {edge_prob!r}.")
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: This check shouldn't be here. `DAG.get_random` should be making these checks.

File: pgmpy/datasets/anm.py [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+        n_nodes : int, optional
+            Number of variables in the generated DAG.
+        edge_prob : float, optional
+            Edge probability for the random DAG.
+        noise : BaseDistribution, optional
+            A ``skpro`` distribution for additive noise.  When
+            ``None``, standard normal :math:`\\mathcal{N}(0, 1)`
+            noise is used.
+        function_type : set, tuple, or list of callable, optional
+            Non-linear functions to randomly apply to each
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: These don't need to be `non-linear` functions. Linear functions work as well.

File: pgmpy/datasets/anm.py [hunk 27]
```diff
@@ -0,0 +1,292 @@
...
+
+            if not parents:
+                data[node] = noise_vals
+            else:
+                parent_data = data[parents].values
+                signal = np.zeros(n)
+                for i, parent_col in enumerate(parents):
+                    w = rng.uniform(*weight_range)
+                    fn = rng.choice(funcs)
+                    signal += w * fn(parent_data[:, i])
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py:143]: This doesn't match the definition that you have written in the class docstring.

File: pgmpy/datasets/anm.py [hunk 28]
```diff
@@ -0,0 +1,292 @@
...
+
+        return DAG.get_random(n_nodes=n_nodes, edge_prob=edge_prob, seed=seed)
+
+    @staticmethod
+    def _sample_noise(
+        noise: Any,
+        n_samples: int,
+        rng: np.random.Generator,
+    ) -> np.ndarray:
+        """Sample additive noise for one node.
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: I don't think we really need a separate method for this. You can keep the logic of this method where it is called from.

File: pgmpy/tests/test_datasets/test_anm.py [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+
+import numpy as np
+import pytest
+from skbase.utils.dependencies import _check_soft_dependencies
+
+from pgmpy.base import DAG
+from pgmpy.datasets import load_dataset
+
+
+def test_load_anm_dataset():
```
Comment (ankurankan) [inline pgmpy/tests/test_datasets/test_anm.py:11]: Add checks for whether correct tags are being set.

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/datasets/anm.py
Comment (ankurankan) [reply pgmpy/datasets/anm.py:1]: @Gitanaskhan26 Please reformat your file to have a 120 char line limit.

File: pgmpy/datasets/anm.py [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    def __init__(
+        self,
+        dag: DAG | None = None,
+        n_nodes: int = 5,
+        edge_prob: float = 0.5,
+        noise: Any = None,
+        function_type: tuple | list = (np.sin, np.cos, np.tanh),
+        weight_range: tuple[float, float] = (-1, 1),
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: I think we should allow this to be either a fixed value or a range. I would suggest renaming this to `weight` and then accepting two possible options a float/int and a tuple of float. If a single value is passed we directly use that otherwise we can sample from the range.

File: pgmpy/datasets/anm.py [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+        self.noise = noise
+        self.function_type = function_type
+        self.weight_range = weight_range
+        self.seed = seed
+
+    def load_dataframe(self, n_samples: int | None = None) -> pd.DataFrame:
+        """Sample data from the generated Additive Noise Model.
+
+        For each child node :math:`X_j`, a function is randomly chosen
+        from *function_type* per parent edge, applied element-wise,
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: I don't think this is the correct behavior. We should be choosing a function for each parent.

File: pgmpy/datasets/anm.py [hunk 32]
```diff
@@ -0,0 +1,150 @@
...
+        return data
+
+    def load_ground_truth(self) -> DAG:
+        """Return the ground-truth DAG of the generated ANM.
+
+        Returns
+        -------
+        DAG
+        """
+        return DAG(self.dag)
```
Comment (ankurankan) [inline pgmpy/datasets/anm.py]: Why not simply return `self.dag`?

File: pgmpy/tests/test_datasets/test_anm.py [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+    ds = load_dataset("anm", seed=42)
+    assert ds.tags["n_variables"] == 5
+    assert ds.tags["n_samples"] == 1000
+    assert ds.tags["is_simulated"] is True
+    assert ds.tags["has_ground_truth"] is True
+    assert ds.tags["is_continuous"] is True
+
+    ds_custom = load_dataset("anm", n_samples=200, seed=7, n_nodes=8, edge_prob=0.3)
+    assert ds_custom.tags["n_variables"] == 8
+    assert ds_custom.tags["n_samples"] == 200
```
Comment (ankurankan) [inline pgmpy/tests/test_datasets/test_anm.py]: Can you remove this test and test for the tag values for all the simulations in `test_load_anm_dataset`

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Review by ankurankan — DISMISSED** [review]
